test/integration/ipamperf/cloud.go

```diff
@@ -0,0 +1,157 @@
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"regexp"
+	"sync"
+
+	"github.com/golang/glog"
+
+	compute "google.golang.org/api/compute/v1"
+	"google.golang.org/api/googleapi"
+	"k8s.io/kubernetes/pkg/controller/nodeipam/ipam/cidrset"
+	"k8s.io/kubernetes/test/integration/util"
+)
+
+var (
+	pathInstance  = regexp.MustCompile(`/compute/(v[\d]+|beta|alpha)/projects/([\w|-]+)/zones/([\w|-]+)/instances/([\w|-]+)(\?.*)?`)
+	pathUpdateNet = regexp.MustCompile(`/compute/(v[\d]+|beta|alpha)/projects/([\w|-]+)/zones/([\w|-]+)/instances/([\w|-]+)/updateNetworkInterface(\?.*)?`)
+	pathOperation = regexp.MustCompile(`/compute/(v[\d]+|beta|alpha)/projects/([\w|-]+)/zones/([\w|-]+)/operations/([\w|-]+)(\?.*)?`)
+)
+
+func writeOperationStatus(w io.Writer, selfLink string, header http.Header, statusCode int, status string) error {
+	return json.NewEncoder(w).Encode(&compute.Operation{
+		ServerResponse: googleapi.ServerResponse{
+			Header:         header,
+			HTTPStatusCode: statusCode,
+		},
+		SelfLink: selfLink,
+		Status:   status,
+	})
+}
+
+type cloudServer struct {
+	serveAt      int              // the qps to server at
+	allocateCIDR bool             // whether cloud should allocate CIDR
+	cidrSet      *cidrset.CidrSet // the cidr set if allocateCIDR is true
+
+	lock      sync.Mutex                   // protect access to the instances map
+	instances map[string]*compute.Instance // known instances
+}
+
+func newCloudServer(serveAt int) *cloudServer {
+	return &cloudServer{
+		serveAt:      serveAt,
+		allocateCIDR: false,
+		instances:    make(map[string]*compute.Instance),
+	}
+}
+
+func newCloudServerWithCIDRAllocator(serveAt int, clusterCIDR *net.IPNet, subnetMaskSize int) *cloudServer {
+	cidrSet, _ := cidrset.NewCIDRSet(clusterCIDR, subnetMaskSize)
+	return &cloudServer{
+		serveAt:      serveAt,
+		allocateCIDR: true,
+		cidrSet:      cidrSet,
+		instances:    make(map[string]*compute.Instance),
+	}
+}
+
+func (cs *cloudServer) getInstanceOrCreate(zone, name string) *compute.Instance {
+	cs.lock.Lock()
+	defer cs.lock.Unlock()
+
+	if instance, found := cs.instances[name]; !found {
+		instance = &compute.Instance{
+			Name: name,
+			Zone: zone,
+			NetworkInterfaces: []*compute.NetworkInterface{
+				{},
+			},
+		}
+		if cs.allocateCIDR {
+			nextRange, _ := cs.cidrSet.AllocateNext()
+			instance.NetworkInterfaces[0].AliasIpRanges = []*compute.AliasIpRange{
+				{
+					IpCidrRange:         nextRange.String(),
+					SubnetworkRangeName: util.TestSecondaryRangeName,
+				},
+			}
+		}
+		cs.instances[name] = instance
+	}
+	return cs.instances[name]
+}
+
+func (cs *cloudServer) assignAliasIPRange(zone, name string, netIf *compute.NetworkInterface) *compute.Instance {
+	instance := cs.getInstanceOrCreate(zone, name)
+
+	cs.lock.Lock()
+	defer cs.lock.Unlock()
+	instance.NetworkInterfaces[0].AliasIpRanges = netIf.AliasIpRanges
+	return instance
+}
+
+func (cs *cloudServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```

Comment:
This looks like a mock for how the GCP API behaves. Is this going to be maintainable?

Reply:
yes, it mocks the "expectation" from the client.